thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java

```diff
@@ -724,13 +735,18 @@ private Response runPreview(long start, long end,
 
       Preconditions.checkNotNull(detectionConfig);
       DetectionPipeline pipeline = this.loader.from(this.provider, detectionConfig, start, end);
-      result = pipeline.run();
-
+      future = this.executor.submit(pipeline::run);
+      result = future.get(PREVIEW_TIMEOUT, TimeUnit.MILLISECONDS);
     } catch (IllegalArgumentException e) {
       return processBadRequestResponse(YamlOperations.PREVIEW.name(), YamlOperations.RUNNING.name(), payload, e);
     } catch (InvocationTargetException e) {
       responseMessage.put("message", "Failed to run the preview due to " + e.getTargetException().getMessage());
       return Response.serverError().entity(responseMessage).build();
+    } catch (TimeoutException e) {
+      // stop the preview
+      future.cancel(true);
```

Comment:
would be safer to move this to a finally block?

Reply:
good idea. updated.